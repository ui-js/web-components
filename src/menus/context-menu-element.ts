import {
    eventLocation,
    keyboardModifiersFromEvent,
    KeyboardModifiers,
    LongPressDetector,
} from '../common/events';
import { UIElement } from '../common/ui-element';
import { MenuItemTemplate } from './menu-core';
import { MENU_TEMPLATE, MENU_STYLE } from './menu-templates';
import { RootMenu } from './root-menu';

/**
 * This web component display a contextual menu when the user performs the
 * appropriate gesture (right-click, control+click, shift+F10, etc...),
 * handles the user interaction to navigate the items in the menu and submenus
 * and either invoke the `onSelect()` hook associated with the selected menu item
 * or dispatches a 'select' event when the user completes their selection.
 *
 * Place this element inside a container and the contextual menu will be
 * displayed when a right-click/control+click/long tap is performed in the
 * container or if the container is keyboard focused when shift+f10 is pressed.
 * In the former case, the menu is diplayed where the click/tap occured.
 * In the later case, the menu is displayed in the center of the parent container.
 *
 * This component is:
 * - screen-reader compatible
 * - keyboard navigable
 * - dark-theme aware
 * - responsive (the menu and submenus will attempt to avoid being displayed
 *   outside of the viewport boundary)
 *
 * Principles of Operation
 *
 * The content of a menu (menu items and submenus) is represented by a 'model',
 * an instance of the Menu class.
 * The model is created from:
 * - argument to the UIContextMenu constructor
 * - setting the `menuItems` property
 * - a `<script>` tag containing a JSON description of menu items
 * - a set of child `<ui-menu-item>` elements.
 *
 * A menu can also have a `<style>` tag, which is applied to style the menu
 * once it is open.
 *
 * The `<ui-context-menu>` and its child elements are kept hidden.
 *
 * When the menu is invoked (with `show()`) a scrim element is created
 * and added as a child of the `<ui-context-menu>`, and a new `<ul>` element
 * is created and attached as a child of the scrim.
 *
 * A set of `<li>` element is added as children of the `<ul>` element, one
 * for each visible menu item, whether this menu item was specified in
 * JSON or with a `<menu-item>` element. The `<li>` element are made of
 * the following components:
 * - a checkmark (optional)
 * - a text label (as a text node, if specified from JSON or as a cloned
 * `UIMenuItemElement` if specified from the `<ui-menu-item>`)
 * - a submenu indicator
 *
 */
export class UIContextMenu extends UIElement {
    private rootMenu: RootMenu;
    // The menu items specified via a constructor or setter
    private templateMenuItems: MenuItemTemplate[];

    private longPressDetector: LongPressDetector;

    constructor(menuItems?: MenuItemTemplate[]) {
        super({
            template: MENU_TEMPLATE,
            style: MENU_STYLE,
        });
        this.templateMenuItems = menuItems ?? [];
    }

    set menuItems(menuItems: MenuItemTemplate[]) {
        this.templateMenuItems = menuItems;
        if (this.rootMenu) {
            this.rootMenu.menuItemTemplates = menuItems;
        }
    }
    get menuItems(): MenuItemTemplate[] {
        return this.templateMenuItems;
    }

    /**
     * @internal
     */
    handleEvent(event: Event): void {
        this.longPressDetector?.dispose();
        this.longPressDetector = undefined;

        if (event.type === 'contextmenu') {
            const evt = event as MouseEvent;
            this.show({
                location: [Math.round(evt.clientX), Math.round(evt.clientY)],
                keyboardModifiers: keyboardModifiersFromEvent(evt),
            });
            event.preventDefault();
            event.stopPropagation();
        } else if (event.type === 'keydown') {
            const evt = event as KeyboardEvent;
            if (
                evt.code === 'ContextMenu' ||
                (evt.code === 'F10' && evt.shiftKey)
            ) {
                // shift+F10 = contextual menu
                // Get the center of the parent
                const bounds = this.parentElement?.getBoundingClientRect();
                if (bounds) {
                    this.show({
                        location: [
                            Math.round(bounds.left + bounds.width / 2),
                            Math.round(bounds.top + bounds.height / 2),
                        ],
                        keyboardModifiers: keyboardModifiersFromEvent(evt),
                    });
                    event.preventDefault();
                    event.stopPropagation();
                }
            }
        } else if (event.type === 'pointerdown') {
            if (event.target === this.shadowRoot.host.parentNode) {
                const pt = eventLocation(event);
                this.longPressDetector = new LongPressDetector(event, () => {
                    this.show({
                        location: pt,
                        keyboardModifiers: keyboardModifiersFromEvent(event),
                    });
                });
                event.preventDefault();
                event.stopPropagation();
            }
        }
    }
    /**
     * Custom elements lifecycle hooks
     * @internal
     */
    connectedCallback(): void {
        super.connectedCallback();
        // Listen for contextual menu in the parent
        const parent = this.parentNode;
        parent.addEventListener('contextmenu', this);
        parent.addEventListener('keydown', this);
        parent.addEventListener('pointerdown', this);
    }
    /**
     * Custom elements lifecycle hooks
     * @internal
     */
    disconnectedCallback(): void {
        super.disconnectedCallback();
        const parent = this.parentNode;
        if (parent) {
            parent.removeEventListener('contextmenu', this);
            parent.removeEventListener('keydown', this);
            parent.removeEventListener('pointerdown', this);
        }
    }
    /**
     * @internal
     */
    focus(): void {
        super.focus();
        if (this.rootMenu?.state !== 'closed') {
            if (this.rootMenu.activeMenuItem) {
                this.rootMenu.activeMenuItem.element.focus();
            } else {
                this.rootMenu.element.focus();
            }
        }
    }
    /**
     * Display the menu at the specified location.
     * If provided, the `keyboardModifiers` option can change what commands
     * are visible, enabled, or what their label is.
     *
     * The contextual menu is shown automatically when the appropriate UI gesture
     * is performed by the user (right-click, shift+F10, etc...). This method
     * only needs to be called to trigger the menu manually (for example to
     * trigger it on click of an item).
     */
    show(options?: {
        location?: [x: number, y: number];
        keyboardModifiers?: KeyboardModifiers;
    }): void {
        if (!this.rootMenu) {
            // Import inline (in the component) style sheet
            this.importStyle();

            // Inline menu items (as a JSON structure in a <script> tag
            // in the markup)
            let jsonMenuItems = this.json;
            if (!Array.isArray(jsonMenuItems)) jsonMenuItems = [];
            this.rootMenu = new RootMenu(
                [...this.templateMenuItems, ...jsonMenuItems],
                {
                    host: this.shadowRoot.host,
                    assignedElement: this.shadowRoot.querySelector<HTMLElement>(
                        'ul'
                    ),
                }
            );
        }
        this.style.display = 'block';
        if (this.rootMenu.show({ ...options, parent: this.shadowRoot })) {
            if (!this.hasAttribute('tabindex')) {
                this.setAttribute('tabindex', '-1');
            }
            this.focus();
        } else {
            this.style.display = 'none';
        }
    }
    /**
     * Hide the menu.
     *
     * The visibility of the menu is typically controlled by the user
     * interaction: the menu is automatically hidden if the user release the
     * mouse button, or after having selected a command. This is a manual
     * override that should be very rarely needed.
     */
    hide(): void {
        this.rootMenu?.hide();
        this.style.display = 'none';
    }
}

export default UIContextMenu;

declare global {
    /** @internal */
    export interface Window {
        UIContextMenu: typeof UIContextMenu;
    }
}
if (!window.customElements.get('ui-context-menu')) {
    window.UIContextMenu = UIContextMenu;
    window.customElements.define('ui-context-menu', UIContextMenu);
}
