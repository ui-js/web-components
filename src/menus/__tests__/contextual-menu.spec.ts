import { UIContextMenu } from '../context-menu-element';

describe('contextual-menu', () => {
  it('renders a simple contextual menu with markup', async () => {
    const element2 = window.document.createElement('div');
    window.document.body.append(element2);
    element2.innerHTML = 'test me';

    const container = window.document.createElement('div');
    window.document.body.append(container);
    container.innerHTML = `<ui-context-menu><script type='application/json'>
        [
            { "label": "Cut"},
            { "label": "Copy"},
            { "label": "Paste"},
            { 
                "label": "Not Visible", 
                "visible": false 
            },
            {
                "label": "Paste as Graphic",
                "disabled": true
            },
            { "type": "divider" },
            {
                "label": "Fence",
                "submenu": [
                    {"label": "Brackets" },
                    {"label": "Square Brackets" },
                    {"label": "Braces" }
                ]
            },
            { "type": "divider" },
            { "label": "Insert Row Before" },
            { "label": "Insert Row After" },
            { "label": "Insert Column Before" },
            { "label": "Insert Column After" }
        ]
    </script>
</ui-context-menu>`;

    const element = window.document.body.getElementsByTagName(
      'ui-context-menu'
    )[0] as UIContextMenu;

    // The following expect() is important.
    // Without it, the './context-menu-element' module get stripped
    // (and therefore the <ui-context-menu> tag fails)
    // as the "as UIContextMenu" statement is not sufficient to
    // keep a reference to the module.
    expect(element instanceof UIContextMenu).toBeTruthy();

    element.show();
    expect(element.shadowRoot?.innerHTML).toMatchSnapshot();
    element.hide();
    expect(element.shadowRoot?.innerHTML).toMatchSnapshot();
  });
});
