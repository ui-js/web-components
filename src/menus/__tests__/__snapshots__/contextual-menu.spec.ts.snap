// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contextual-menu renders a simple contextual menu with markup 1`] = `
"<slot></slot><style>
*,
::before,
::after {
    box-sizing: border-box;
}
:host {
    display: none;
    color-scheme: light dark;
    -webkit-user-select: none;  /* Important: Safari iOS doesn't respect user-select */
    user-select: none;
    cursor: default;
    -webkit-touch-callout: none;
    -webkit-tap-highlight-color: rgba(0 0 0 0);
    --active-label-color: #fff;
    --label-color: #121212;
    --menu-bg: #e2e2e2;
    --active-bg: #5898ff;
    --active-bg-dimmed: #c5c5c5;
}
:host([hidden]) {
    display: none;
}
:host([disabled]) {
    pointer-events: none;
    opacity:  .5;
}
:host(:focus), :host(:focus-within) {
    outline: Highlight auto 1px;    /* For Firefox */
    outline: -webkit-focus-ring-color auto 1px;
}
:host div.scrim {
    position: fixed;
    contain: content;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    z-index: 9999;
    outline: none;
    background: transparent;
}
:host slot {
    display: none;
}
ul.menu-container {
    position: absolute;
    width: auto;
    height: auto;
    z-index: 10000;
    border-radius: 8px;
    background: var(--menu-bg);
    box-shadow: 0 0 2px rgba(0, 0, 0, .5), 0 0 20px rgba(0, 0, 0, .2);

    list-style: none;
    padding: 6px 0 6px 0;
    margin: 0;
    user-select: none;
    cursor: default;

    color: var(--label-color);
    font-weight: normal;
    font-style: normal;
    text-shadow: none;
    text-transform: none;
    letter-spacing: 0;
    outline: none;
    opacity: 1;
}
ul.menu-container > li {
    display: flex;
    flex-flow: row;
    align-items: center;
    padding: 1px 7px 1px 7px;
    margin-top: 0;
    margin-left: 6px;
    margin-right: 6px;
    border-radius: 4px;
    white-space: nowrap;
    position: relative;
    outline: none;
    fill: currentColor;
    user-select: none;
    cursor: default;
    text-align: left;
    color: inherit;

    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
    font-size: 13px;
    line-height: 16px;
    letter-spacing: 0.007em;
}
ul.menu-container > li > .label {
    appearance: none;
    background: none;
    outline: none;
    width: 100%;
    margin: 0;
    padding: 1px 2px 1px 1px;
    overflow: visible;
    border: 1px solid transparent;
    white-space: nowrap;
}

ul.menu-container > li > .label.indent {
    margin-left: 12px;
}
ul.menu-container > li[role=separator] {
    border-bottom: 1px solid #c7c7c7;
    border-radius: 0;
    padding: 0;
    margin-left: 15px;
    margin-right: 15px;
    padding-top: 5px;
    margin-bottom: 5px;
    width: calc(100% - 30px);
}
ul.menu-container > li[aria-disabled=true] {
    opacity: .5;
}

ul.menu-container > li.active {
    background: var(--active-bg);
    background: -apple-system-control-accent;
    color: var(--active-label-color);
}

ul.menu-container > li.active.is-submenu-open {
    background: var(--active-bg-dimmed);
    color: inherit;
}

ul.menu-container > li[aria-haspopup=true]>.label {
     padding-right: 0;
}

.right-chevron {
    margin-left: 24px;
    width: 10px;
    height: 10px;
    margin-bottom: 4px;
}
.checkmark {
    margin-right: -11px;
    margin-left: -4px;
    margin-top : 2px;
    width: 16px;
    height: 16px;
}

ul.menu-container > li[aria-haspopup=true].active::after {
    color: white;
}
@media (prefers-color-scheme: dark) {
    :host {
        --label-color: #fff;
        --active-label-color: #000;
        --menu-bg: #525252;
        --active-bg: #5898ff;
        --active-bg-dimmed: #5c5c5;
    }
}
</style><div role=\\"presentation\\" style=\\"position: fixed; contain: content; top: 0px; left: 0px; right: 0px; bottom: 0px; z-index: 9999; outline: none; background: transparent;\\"><ul class=\\"menu-container\\" part=\\"menu-container\\" tabindex=\\"-1\\" role=\\"menu\\" aria-orientation=\\"vertical\\"><li part=\\"menu-item\\" tabindex=\\"0\\" role=\\"menuitem\\"><span class=\\"label\\">Cut</span></li><li part=\\"menu-item\\" tabindex=\\"-1\\" role=\\"menuitem\\"><span class=\\"label\\">Copy</span></li><li part=\\"menu-item\\" tabindex=\\"-1\\" role=\\"menuitem\\"><span class=\\"label\\">Paste</span></li><li part=\\"menu-item\\" tabindex=\\"-1\\" role=\\"menuitem\\"><span class=\\"label\\">Not Visible</span></li><li part=\\"menu-item\\" tabindex=\\"-1\\" role=\\"menuitem\\" aria-disabled=\\"true\\"><span class=\\"label\\">Paste as Graphic</span></li><li part=\\"menu-separator\\" role=\\"separator\\"></li><li part=\\"menu-item\\" tabindex=\\"-1\\" role=\\"menuitem\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\"><span class=\\"label\\">Fence</span><span aria-hidden=\\"true\\" class=\\"right-chevron\\"><svg focusable=\\"false\\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 320 512\\"><path fill=\\"currentColor\\" d=\\"M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z\\"></path></svg></span></li><li part=\\"menu-separator\\" role=\\"separator\\"></li><li part=\\"menu-item\\" tabindex=\\"-1\\" role=\\"menuitem\\"><span class=\\"label\\">Insert Row Before</span></li><li part=\\"menu-item\\" tabindex=\\"-1\\" role=\\"menuitem\\"><span class=\\"label\\">Insert Row After</span></li><li part=\\"menu-item\\" tabindex=\\"-1\\" role=\\"menuitem\\"><span class=\\"label\\">Insert Column Before</span></li><li part=\\"menu-item\\" tabindex=\\"-1\\" role=\\"menuitem\\"><span class=\\"label\\">Insert Column After</span></li></ul></div>"
`;

exports[`contextual-menu renders a simple contextual menu with markup 2`] = `
"<slot></slot><style>
*,
::before,
::after {
    box-sizing: border-box;
}
:host {
    display: none;
    color-scheme: light dark;
    -webkit-user-select: none;  /* Important: Safari iOS doesn't respect user-select */
    user-select: none;
    cursor: default;
    -webkit-touch-callout: none;
    -webkit-tap-highlight-color: rgba(0 0 0 0);
    --active-label-color: #fff;
    --label-color: #121212;
    --menu-bg: #e2e2e2;
    --active-bg: #5898ff;
    --active-bg-dimmed: #c5c5c5;
}
:host([hidden]) {
    display: none;
}
:host([disabled]) {
    pointer-events: none;
    opacity:  .5;
}
:host(:focus), :host(:focus-within) {
    outline: Highlight auto 1px;    /* For Firefox */
    outline: -webkit-focus-ring-color auto 1px;
}
:host div.scrim {
    position: fixed;
    contain: content;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    z-index: 9999;
    outline: none;
    background: transparent;
}
:host slot {
    display: none;
}
ul.menu-container {
    position: absolute;
    width: auto;
    height: auto;
    z-index: 10000;
    border-radius: 8px;
    background: var(--menu-bg);
    box-shadow: 0 0 2px rgba(0, 0, 0, .5), 0 0 20px rgba(0, 0, 0, .2);

    list-style: none;
    padding: 6px 0 6px 0;
    margin: 0;
    user-select: none;
    cursor: default;

    color: var(--label-color);
    font-weight: normal;
    font-style: normal;
    text-shadow: none;
    text-transform: none;
    letter-spacing: 0;
    outline: none;
    opacity: 1;
}
ul.menu-container > li {
    display: flex;
    flex-flow: row;
    align-items: center;
    padding: 1px 7px 1px 7px;
    margin-top: 0;
    margin-left: 6px;
    margin-right: 6px;
    border-radius: 4px;
    white-space: nowrap;
    position: relative;
    outline: none;
    fill: currentColor;
    user-select: none;
    cursor: default;
    text-align: left;
    color: inherit;

    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
    font-size: 13px;
    line-height: 16px;
    letter-spacing: 0.007em;
}
ul.menu-container > li > .label {
    appearance: none;
    background: none;
    outline: none;
    width: 100%;
    margin: 0;
    padding: 1px 2px 1px 1px;
    overflow: visible;
    border: 1px solid transparent;
    white-space: nowrap;
}

ul.menu-container > li > .label.indent {
    margin-left: 12px;
}
ul.menu-container > li[role=separator] {
    border-bottom: 1px solid #c7c7c7;
    border-radius: 0;
    padding: 0;
    margin-left: 15px;
    margin-right: 15px;
    padding-top: 5px;
    margin-bottom: 5px;
    width: calc(100% - 30px);
}
ul.menu-container > li[aria-disabled=true] {
    opacity: .5;
}

ul.menu-container > li.active {
    background: var(--active-bg);
    background: -apple-system-control-accent;
    color: var(--active-label-color);
}

ul.menu-container > li.active.is-submenu-open {
    background: var(--active-bg-dimmed);
    color: inherit;
}

ul.menu-container > li[aria-haspopup=true]>.label {
     padding-right: 0;
}

.right-chevron {
    margin-left: 24px;
    width: 10px;
    height: 10px;
    margin-bottom: 4px;
}
.checkmark {
    margin-right: -11px;
    margin-left: -4px;
    margin-top : 2px;
    width: 16px;
    height: 16px;
}

ul.menu-container > li[aria-haspopup=true].active::after {
    color: white;
}
@media (prefers-color-scheme: dark) {
    :host {
        --label-color: #fff;
        --active-label-color: #000;
        --menu-bg: #525252;
        --active-bg: #5898ff;
        --active-bg-dimmed: #5c5c5;
    }
}
</style>"
`;
